platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/collections/LongListDisk.java

```diff
@@ -40,29 +90,103 @@ public class LongListDisk extends LongList {
      * @throws IOException If there was a problem reading the file
      */
     public LongListDisk(final Path file) throws IOException {
-        super(FileChannel.open(file, StandardOpenOption.CREATE, StandardOpenOption.READ, StandardOpenOption.WRITE));
-        this.file = file;
+        this(file, DEFAULT_RESERVED_BUFFER_LENGTH);
+    }
+
+    LongListDisk(final Path file, final long reservedBufferLength) throws IOException {
+        super(file, reservedBufferLength);
+        freeChunks = new ConcurrentLinkedDeque<>();
+    }
+
+    @Override
+    protected void readBodyFromFileChannelOnInit(String sourceFileName, FileChannel fileChannel) throws IOException {
+        Path tempFile = createTempFile(sourceFileName);
+        // create temporary file for writing
+        @SuppressWarnings("resource")
+        final RandomAccessFile rf = new RandomAccessFile(tempFile.toFile(), "rw");
+        // ensure that the amount of disk space is enough
+        // two additional chunks are required to accommodate "compressed" first and last chunks in the original file
+        rf.setLength(fileChannel.size() + (long) 2 * memoryChunkSize);
+        currentFileChannel = rf.getChannel();
+
+        final int totalNumberOfChunks = calculateNumberOfChunks(size());
+        final int firstChunkWithDataIndex = (int) (minValidIndex.get() / numLongsPerChunk);
+        final int minValidIndexInChunk = (int) (minValidIndex.get() % numLongsPerChunk);
+
+        // copy the first chunk
+        final ByteBuffer transferBuffer = initOrGetTransferBuffer();
+        transferBuffer.position(minValidIndexInChunk * Long.BYTES);
+        MerkleDbFileUtils.completelyRead(fileChannel, transferBuffer);
+        transferBuffer.flip();
+        // writing the full chunk, all values before minValidIndexInChunk are zeroes
+        MerkleDbFileUtils.completelyWrite(currentFileChannel, transferBuffer, 0);
+        chunkList.set(firstChunkWithDataIndex, 0L);
+
+        // copy everything except for the first chunk and the last chunk
+        MerkleDbFileUtils.completelyTransferFrom(
+                currentFileChannel, fileChannel, memoryChunkSize, (long) (totalNumberOfChunks - 2) * memoryChunkSize);
+
+        // copy the last chunk
+        transferBuffer.clear();
+        MerkleDbFileUtils.completelyRead(fileChannel, transferBuffer);
+        transferBuffer.flip();
+        MerkleDbFileUtils.completelyWrite(
+                currentFileChannel, transferBuffer, (long) (totalNumberOfChunks - 1) * memoryChunkSize);
+
+        for (int i = firstChunkWithDataIndex + 1; i < totalNumberOfChunks; i++) {
+            chunkList.set(i, (long) (i - firstChunkWithDataIndex) * memoryChunkSize);
+        }
+    }
+
+    private ByteBuffer initOrGetTransferBuffer() {
+        final ByteBuffer buffer;
+        if (TRANSFER_BUFFER_THREAD_LOCAL.get() == null) {
```

Comment:
Do `get()` once:
```
final ByteBuffer buffer = TRANSFER_BUFFER_THREAD_LOCAL.get();
if (buffer == null) {
...
```

Reply:
cannot be final due to allocate on 144, but still a good change without final.

Reply:
Fixed